rtmr/extend/extend.go

```diff
@@ -0,0 +1,199 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package extend
+
+import (
+	"bytes"
+	"crypto"
+	"crypto/sha512" // Registrer SHA384 and SHA512
+	"fmt"
+	"os"
+	"path/filepath"
+	"strconv"
+
+	"github.com/google/go-configfs-tsm/configfs/configfsi"
+	"github.com/google/go-configfs-tsm/configfs/linuxtsm"
+	"github.com/google/logger"
+)
+
+const (
+	RtmrSubsystem = "rtmrs"
+	TsmRtmrPrefix = configfsi.TsmPrefix + "/" + RtmrSubsystem
+	TsmRtmrDigest = "digest"
+	TsmPathIndex  = "index"
+	TsmPathTcgMap = "tcg_map"
+)
+
+// RtmrExtend is a struct that represents a rtmr entry in the configfs.
+type RtmrExtend struct {
+	RtmrIndex int
+	entry     *configfsi.TsmPath
+	client    configfsi.Client
+}
+
+// Remove the null bytes from the c-string
+func convertCstring(b []byte) string {
+	return string(bytes.Trim(b, "\x00"))
+}
+
+func (r *RtmrExtend) attribute(subtree string) string {
+	a := *r.entry
+	a.Attribute = subtree
+	return a.String()
+}
+
+// ExtendDigest extends the measurement to the rtmr with the given hash.
+func (r *RtmrExtend) ExtendDigest(hash []byte) error {
+	if (r.RtmrIndex != 2) && (r.RtmrIndex != 3) {
+		return fmt.Errorf("invalid rtmr index %d, userspace can only extend to rtmr2 or rtmr3", r.RtmrIndex)
+	}
+	if err := r.client.WriteFile(r.attribute(TsmRtmrDigest), hash); err != nil {
+		return fmt.Errorf("could not write report %s: %v", TsmPathIndex, err)
+	}
+	return nil
+}
+
+// SetRtmrIndex sets a configfs rtmr entry to the given index.
+// It reports an error if the index cannot be written or the rtmr_tcg map does not match.
+func (r *RtmrExtend) SetRtmrIndex() error {
+	indexBytes := []byte(strconv.Itoa(r.RtmrIndex)) // Convert index to []byte
+	indexPath := r.attribute(TsmPathIndex)
+	if err := r.client.WriteFile(indexPath, indexBytes); err != nil {
+		return fmt.Errorf("could not write index %s: %v", indexPath, err)
+	}
+	if err := r.ValidateRtmrIndex(); err != nil {
+		return fmt.Errorf("the tcg_map is invalid %s: %v", indexPath, err)
+	}
+	return nil
+}
+
+// ValidateRtmrIndex checks if the rtmr to PCR map is valid.
+// It returns an error if the rtmr_tcg map does not match the expected value.
+func (r *RtmrExtend) ValidateRtmrIndex() error {
+	if (r.client == nil) || (r.entry == nil) {
+		return fmt.Errorf("RtmrExtend is not initialized")
+	}
+	data, err := r.client.ReadFile(r.attribute(TsmPathTcgMap))
+	if err != nil {
+		return fmt.Errorf("could not read  %s: %v", TsmPathTcgMap, err)
+	}
+	var rtmrPcrMaps = map[int]string{
+		0: "1,7\n",
+		1: "2-6\n",
+		2: "8-15\n",
+		3: "\n",
+	}
+
+	inputMap := convertCstring(data)
+	if inputMap != rtmrPcrMaps[r.RtmrIndex] {
+		return fmt.Errorf("tcg map error. expect:%s get:%s", rtmrPcrMaps[r.RtmrIndex], string(data))
+	}
+	return nil
+}
+
+// SearchRtmrInterface searches for an rtmr entry in the configfs.
+func SearchRtmrInterface(client configfsi.Client, index int) (*RtmrExtend, error) {
```

Comment:
Why search instead of creating an ephemeral tsm entry per request? That's more in line with what go-configfs-tsm does with reports. Avoids potential data races.

Reply:
The RTMR tsm subsystem is different from the report one. 
For one index, only once entry (Interface) is allowed. https://github.com/intel/tdx/blob/8c7a50a7ecf686232231ba15e55d9d39876df4c9/drivers/virt/coco/tsm.c#L437

If someone has created an entry for rtmr1, we can not create a new entry for rtmr1.